### ISSUE TITLE
Major code refactoring and modularization

### ISSUE THREAD

**Issue by pofider** [jsreport/jsreport#140]
I would like to split this repository and make jsreport sources more modularized. There are currently 28 default extensions and many of them are just increasing the package size. Also the code quality goes rapidly down with such a "huge" codebase. There are many other obvious pros I don't need to mention so just lets do it!

Topics to discuss....

## What should be in the core
I would like to extract every extension to the separate module. The question is what should stay in the "jsreport-core" package. I would propose that "jsreport-core" would just contain the extensions management, configuration management and the main Reporter class. So lightweight, stateless, without rest API but extensible. It wouldn't be basically useful without additionally installed extensions but you get only what you ask for. 

## Repository names
This jsreport repository would then contain just the previously mentioned core implementation. We would only additionally provide an easy to use npm package "jsreport-full" with the included dependencies to major extensions.

## UI code

Many extensions includes jsreport studio front-end part. For example `scheduling` extension not only runs background rendering jobs but also provides UI to define these jobs. This is not conceptually perfect, I know. The question is if it would make sense to split this into packages like 

`jsreport-scripts`, `jsreport-scripts-studio`, `jsreport-jade`, `jsreport-jade-studio`.....

Also the `express` extension would be then split into the extension `jsreport-api` and `jsreport-studio` .

Isn't this grained to much? Isn't the studio just an integral part and should be treated as such? Or maybe it should depend case by case?

## Coding standards
I would like to take a look at the extensions coding styles after separating them into the dedicated repositories. We should agree on the coding styles on the one particular extension and then make the others in the similar way. We should use ES6, write the tests in the similar way, do the linting and so on. Of course, every extension can be written in the different way and it is up to the author, however we will need to maintain more than 40 repositories for extensions so it would be nice to have them unified.

What do you think about the eslint [feross/standard](https://github.com/feross/standard) ? I personally like to use semicolons at the end of the statements, this is kind of weird about this package, but I would be fine with it in the end.

**Comment by bjrmatos** [comment]
> I would like to split this repository and make jsreport sources more modularized. There are currently 28 default extensions and many of them are just increasing the package size. Also the code quality goes rapidly down with such a "huge" codebase. There are many other obvious pros I don't need to mention so just lets do it!
> 
> Topics to discuss....

totally agree with you!! lets do it :tada: 



> I would propose that "jsreport-core" would just contain the extensions management, configuration management and the main Reporter class. So lightweight, stateless, without rest API but extensible. It wouldn't be basically useful without additionally installed extensions but you get only what you ask for.

this is the way that webpack and babel (with the new v6) works (a core package, and plugins, extensions on its own packages), so for me this is the way to do it too



> This jsreport repository would then contain just the previously mentioned core implementation. We would only additionally provide an easy to use npm package "jsreport-full" with the included dependencies to major extensions.

so basically the current version of jsreport will be jsreport-full? extension presets are very useful (again, something like what [babel is doing](https://www.npmjs.com/package/babel-preset-es2015))



> UI code
> 
> Many extensions includes jsreport studio front-end part. For example scheduling extension not only runs background rendering jobs but also provides UI to define these jobs. This is not conceptually perfect, I know. The question is if it would make sense to split this into packages like
> 
> jsreport-scripts, jsreport-scripts-studio, jsreport-jade, jsreport-jade-studio.....
> 
> Also the express extension would be then split into the extension jsreport-api and jsreport-studio .
> 
> Isn't this grained to much? Isn't the studio just an integral part and should be treated as such? Or maybe it should depend case by case?

i think the studio is just an integral part (for now), if there is **so much interesting** in the community to implement pluggeable UI, well it is something we can do it in the future.. i think we should concentrate in other tasks (we have many work to do..), also you have proved a way to implement custom syntax highlighting for templates engines, i think that's enough for now, we can extract the studio/UI part of packages later (if necessary)






> Coding standards
> 
> I would like to take a look at the extensions coding styles after separating them into the dedicated repositories. We should agree on the coding styles on the one particular extension and then make the others in the similar way. We should use ES6, write the tests in the similar way, do the linting and so on. Of course, every extension can be written in the different way and it is up to the author, however we will need to maintain more than 40 repositories for extensions so it would be nice to have them unified.
> 
> What do you think about the eslint feross/standard ? I personally like to use semicolons at the end of the statements, this is kind of weird about this package, but I would be fine with it in the end.

 i also like (and i think it makes code more readable) semicolons, so if we are going to look in the `standard` family we can use [semistandard](https://github.com/Flet/semistandard) (the same standard's rules but with semicolons). 

other options are [eslint-config-airbnb](https://github.com/airbnb/javascript/tree/master/packages/eslint-config-airbnb), it has rules for es6, and react development, i use that on my projects and apps in my work (in the end of the day i still [customize the rules a little bit](https://github.com/bjrmatos/electron-workers/blob/master/.eslintrc) to match my coding style)

**Comment by bjrmatos** [comment]
also i suggest the use of [greenkeeper](http://greenkeeper.io/) to keep it up to date with new versions of dependencies.. it is [really cool!!](https://github.com/bjrmatos/electron-workers/pull/2)

**Comment by pofider** [comment]
I have done some progress here and extracted all of the extensions into the separate packages ([jsreport-scripts](https://github.com/jsreport/jsreport-scripts),  [jsreport-phantom-pdf](https://github.com/jsreport/jsreport-phantom-pdf)...). I use eslint standard without semicolons, in the end I came to conclusion it is not a big deal anyway and better to jump in to the more popular stream. There is still a lot of work before we can release it - documentation, refactoring, tests... but it is on the way. I am looking forward to make this done and start realizing all the other suggestions quicker.

Now I am about to delete the extensions from the jsreport repository and shape it up to the first pre-release. However I am still struggling my mind if the main repository `jsreport` should be the repository with the minimalist core or the one with full set. There are basically two ways:

**a) jsreport repository is the full set as it is right now**
The core minimalist code is extracted into the separate repository jsreport-core. nodejs developers can use the minimalist jsreport-core together with additional extensions. Other developers just install through the jsreport package.
This is back compatible, jsreport brand (package) still represents the full official distribution. 

**b) jsreport repository is the actual minimalist core**
The full distribution is extracted into package, don't know the name yet. Something like jsreport-studio / jsreport-full / jsreport-dist .... Such a repository would basically just contain the package.json with all of the main extensions in the dependencies. nodejs developers would use jsreport package together with additional extensions. Other developers would install through jsreport-full package.
This not back-compatible and breaks for all existing implementations.

This is important decision, what you think? I am changing my mind every day and would love to have an external opinion on this particular thing.

**Comment by bjrmatos** [comment]
> I use eslint standard without semicolons, in the end I came to conclusion it is not a big deal anyway and better to jump in to the more popular stream.

well i can confess now that i hate `standard` rules hahah but whatever, maybe is a good chance to try a new style..



> The core minimalist code is extracted into the separate repository jsreport-core. nodejs developers can use the minimalist jsreport-core together with additional extensions. Other developers just install through the jsreport package.
> This is back compatible, jsreport brand (package) still represents the full official distribution.

the big thing here is **"backwards compatible"**, i think the right move is keep `jsreport` repository as the full version (just like now) and making it depend on `jsreport-core`, also keeping jsreport package as the "ready to use" version will make it looks like if nothing had happened internally

**Comment by bjrmatos** [comment]
question: i have noticed that the package name of [jsreport-scripts](https://github.com/jsreport/jsreport-scripts) repository is [jsreport-analytics](https://github.com/jsreport/jsreport-scripts/blob/master/package.json#L2), was this intentional?

**Comment by pofider** [comment]
>  i think the right move is keep jsreport repository as the full version (just like now) and making it depend on jsreport-core

Ok. I will make it so.

> question: i have noticed that the package name of jsreport-scripts repository is jsreport-analytics, was this intentional?

No, it is just unfinished. Thanks for pointing it out.

**Comment by pofider** [comment]
So I extracted the jsreport core into the separate repository and pushed the changes to jsreport git master. It should be already working and back compatible. However there is still plenty of work to be done before I am satisfied with it...

I will continue on this so we get ASAP to the release point also with the new default fs-store. After that we should be able to introduce new super cool extensions and recipes every day :)

**Comment by bjrmatos** [comment]
:tada: good news :tada: 

![giphy](https://cloud.githubusercontent.com/assets/4262050/11165337/f88d9214-8ad9-11e5-98b4-8556a0c47048.gif)

**Comment by pofider** [comment]
Small step for npm, huge step for jsreport.

I have published jsreport 0.10 which includes this major code refactoring. It has also fs store set as default. And all these changes are still 100% back compatible. Great!.

I am thinking now that it is probably time to end the jsreport <-> toner schism. Because jsreport-core now has exactly the same function as toner. It is probably time to merge toner to jsreport-core and document out the usage of jsreport-core.

**Comment by bjrmatos** [comment]
> Small step for npm, huge step for jsreport.
> 
> I have published jsreport 0.10 which includes this major code refactoring. It has also fs store set as default. And all these changes are still 100% back compatible. Great!.

wonderful job! 



> I am thinking now that it is probably time to end the jsreport <-> toner schism. Because jsreport-core now has exactly the same function as toner. It is probably time to merge toner to jsreport-core and document out the usage of jsreport-core.

the name for the `toner-*` modules, will still start with the `toner` prefix? since those modules are render only logic (without jsreport integration) i think would make sense to keep the `toner` prefix. or do you have a better name/idea for those modules?

**Comment by pofider** [comment]
I was basically thinking to take the code from toner-jsrender, toner-phantom-pdf and put it to jsreport-jsrender and jsreport-phantom-pdf respectively. 

Motivation and problems
- we now maintain two parallel libraries for rendering
- it is complicated looking at jsreport-jsrender which is referencing toner-jsrender
- we always need to release like three packages now to get the fix to jsreport
- original motivation for toner is covered now in jsreport-core. It is lightweight and free
- toner repository is not active, if we forward everybody to jsreport-core, it could get very live

What are the consequences for end user?
- jsreport-core will provide very similar API to toner
- jsreport-core is almost the same lightweight as toner
- packages like jsreport-jsrender are identical to toner-jsrender
- packages like jsreport-phantom-pdf contains a little bit of additional code for extending ui and entity model, but this is applied just conditionally and if it is too much we can always split it into two jsreport packages

Toner can stay like it is, but I think people will anyway prefer jsreport-core.

**Comment by bjrmatos** [comment]
i agree with you, if we maintain the same benefits of toner then i don't see any problem, jsporte-core FTW!

**Comment by pofider** [comment]
Done! It is now reality.
